cmd/goplugin-heroku/resource/build.go

```diff
@@ -0,0 +1,256 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-getter/helper/url"
+	"github.com/hashicorp/go-hclog"
+	heroku "github.com/heroku/heroku-go/v5"
+)
+
+type Build struct {
+	AppID      string
+	BuildID    *string
+	Buildpacks BuildPacks `puppet:"value=>[]"`
+	SourceBlob SourceBlob
+}
+
+type BuildPack struct {
+	Name *string
+	URL  *string
+}
+
+type BuildPacks []*BuildPack
+
+type SourceBlob struct {
+	Checksum *string
+	URL      *string
+	Version  *string
+}
+
+type BuildHandler struct{}
+
+func (*BuildHandler) Create(desiredState *Build) (*Build, string, error) {
+	hclog.Default().Debug("Creating Heroku Build", "desiredState", desiredState)
+
+	fmt.Println(convertToBuildOpts(desiredState))
+
+	s := herokuService()
+
+	buildOpts, err := convertToBuildOpts(desiredState)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Creates a Heroku Build and allocates an ID that can be used to read it in the future
+	build, err := s.BuildCreate(context.TODO(), desiredState.AppID, *buildOpts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	desiredState.BuildID = &build.ID
+	desiredState.AppID = build.App.ID
+
+	return desiredState, fmt.Sprintf("herokuid:/%s?build_id=%s&url=%s", build.App.ID, build.ID, *desiredState.SourceBlob.URL), nil
+}
+
+func (*BuildHandler) Read(externalID string) (*Build, error) {
+	hclog.Default().Debug("Reading Heroku Build", "externalID", externalID)
+
+	s := herokuService()
+
+	extUrl, err := url.Parse(externalID)
+	if err != nil {
+		return nil, err
+	}
+
+	extAppID := extUrl.Path[1:]
+	extBuildID := extUrl.Query().Get(`build_id`)
+	extBuildURL := extUrl.Query().Get(`url`)
+	build, err := s.BuildInfo(context.TODO(), extAppID, extBuildID)
+	if err != nil {
+		return nil, err
+	}
+
+	build.SourceBlob.URL = extBuildURL
+
+	actualState, err := convertToBuild(build)
+	if err != nil {
+		return nil, err
+	}
+
+	return actualState, nil
+}
+
+func (*BuildHandler) Update(externalID string, desiredState *Build) (*Build, error) {
+	hclog.Default().Debug("Updating Build", "desiredState", desiredState)
+
+	s := herokuService()
+
+	extUrl, err := url.Parse(externalID)
+	if err != nil {
+		return nil, err
+	}
+
+	appId := extUrl.Path[1:] // Strip leading '/'
+
+	buildOpts, err := convertToBuildOpts(desiredState)
+	if err != nil {
+		return nil, err
+	}
+
+	actualState, err := s.BuildCreate(context.TODO(), appId, *buildOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	desiredState, err = convertToBuild(actualState)
+	if err != nil {
+		return nil, err
+	}
+
+	return desiredState, nil
+}
+
+func (*BuildHandler) Delete(externalID string) error {
+	hclog.Default().Debug("Build cannot be deleted", "externalID", externalID)
```

Comment:
Both the name `build` and the fact that it cannot be deleted makes me wonder if a build should be represented as a resource step. From reading the docs it looks more like an action. Perhaps you should instead define a Go action that is callable from the lyra workflow? See for instance "workflows/called_go.ll" and "examples/go-samples/called.go" for a simple example of how to serve up one action. An alternative is to let your plugin provide the action. You would do something like this in your `Server` function:
```Go
_, file, line, _ := runtime.Caller(0)
loc := issue.NewLocation(file, line + 3, 0) // needed for correct for error reporting. Should appoint start of Do code-block
a := lyra.Action{
  Do: func(build *Build) *SourceBlob {
    // Run build here
  }
}
sb.RegisterStep(a.Resolve(c, "heroku::build", loc))
```

Reply:
Is there a way to build a Lyra Go action into the go-plugin itself without requiring users to define the action themselves? That's what I'm trying to avoid.

Reply:
Yes, that's the alternative I describe at the end of my previous comment. It goes into your `Server` function.

Reply:
That's possible, it would require some additional fields be available for the app itself as we'd need to know when the build was going to change and trigger a new build if changes were made

Reply:
After running some tests with the heroku builds API, I've found that the API is not idempotent. With that in mind, leaving Build as a resource for now will probably be necessary until a solution can be found to check and compare build state so that a user doesn't trigger a new build everything they run Lyra.